### PR TITLE
Feature ge scheduling basic and complex renovation

### PR DESCRIPTION
**Pull request by dbulaja98**
In this PR, we have implemented basic and complex renovation. Both basic and complex renovations last all day, we do not entry time. Basic renovation is only about scheduling room in given time interval, while complex renovation has two types: divide and join.  In join we pick second room, which must be next to primary room. After scheduling both rooms will be unavailable in given time interval. Divide is much like basic renovation, but there will be some difference later.

### CODE REVIEW

**Comment by rammba** [comment]
[(0) PR Feature ge scheduling basic and complex renovation](https://trello.com/c/NGjfZ1BW/145-0-pr-feature-ge-scheduling-basic-and-complex-renovation)

File: OcelotApiGateway/ocelot.global.json [hunk 1]
```diff
@@ -278,16 +278,42 @@
...
+=======
     {
       "DownstreamPathTemplate": "/schedule/examination/by-specialty/{everything}",
+>>>>>>> master
+      "DownstreamScheme": "http",
+      "DownstreamHostAndPorts": [
+        {
+          "Host": "$PSW_SCHEDULE_SERVICE_HOST",
+<<<<<<< HEAD
+          "Port": "$PSW_SCHEDULE_SERVICE_PORT"
```
Comment (rammba) [inline OcelotApiGateway/ocelot.global.json:294]: Resolve these conflicts and remove quotes for port env var.

File: Schedule.API/Services/Procedures/RecommendationService.cs [hunk 2]
```diff
@@ -253,5 +254,94 @@ private IEnumerable<EquipmentRelocationDto> FindCommonEquipmentRelocationAppoint
...
+
+        private IEnumerable<RoomRenovationDto> BasicRenovationSearch(SchedulingDto dto)
+        {
+            List<RoomRenovationDto> appointments = new List<RoomRenovationDto>();
+            List<Shift> availableShiftsSource = new List<Shift>();
+            IEnumerable<Shift> shiftsForSourceRoom = _shiftWrapper.Repository.GetShiftsByRoomID(dto.FirstRoomId);
+            foreach (Shift shift in shiftsForSourceRoom)
+            {
+                List<Shift> shiftList = new List<Shift>();
+                shiftList.Add(shift);
```
Comment (rammba) [inline Schedule.API/Services/Procedures/RecommendationService.cs:274]: This list will always have only one shift. Probably need to be defined before for.

File: Schedule.API/Services/Procedures/RecommendationService.cs [hunk 3]
```diff
@@ -253,5 +254,94 @@ private IEnumerable<EquipmentRelocationDto> FindCommonEquipmentRelocationAppoint
...
+                IEnumerable<RecommendationDto> recommendationsSourceRoom = FindRecommendationsInShifts(shiftList, 16);
+                if (recommendationsSourceRoom.Count() == 16)
+                {
+                    availableShiftsSource.Add(shift);
+                }
+            }
+            foreach (Shift shift in shiftsForDestinationRoom)
+            {
+                List<Shift> shiftList = new List<Shift>();
+                shiftList.Add(shift);
```
Comment (rammba) [inline Schedule.API/Services/Procedures/RecommendationService.cs:315]: Same as above.

File: Schedule.API/DTOs/RoomRenovationDto.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using Schedule.API.Model.Utilities;
+
+namespace Schedule.API.DTOs
+{
+    public class RoomRenovationDto
+    {
+        public int FirstRoomId { get; set; }
+        public int SecondRoomId { get; set; }
```
Comment (BoKna031) [inline Schedule.API/DTOs/RoomRenovationDto.cs:8]: Better name them Source and Destination Room

File: Schedule.API/DTOs/SchedulingDto.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using Schedule.API.Model.Utilities;
+
+namespace Schedule.API.DTOs
+{
+    public class SchedulingDto
+    {
+        public int FirstRoomId { get; set; }
+        public int SecondRoomId { get; set; }
```
Comment (BoKna031) [inline Schedule.API/DTOs/SchedulingDto.cs:8]: Better name them Source and Destination Room

File: WPFHospitalEditor/SearchResultDialog.xaml.cs [hunk 6]
```diff
@@ -169,7 +189,29 @@ private void AddScheduleRelocationButton(int row, EquipmentRelocationDto relocat
...
+
+            scheduleBtn.Click += (s, e) =>
+            {
+                List<int> doctors = doctorServerController.GetDoctorsByRoomsAndShifts(schDto).ToList();
+                foreach (int doctorId in doctors)
+                {
+                    renovationServerController.ScheduleRenovation(startDate, endDate, doctorId, AllConstants.PatientIdForRenovation);
+                }
+                MessageBox.Show("Renovation is successfully scheduled!", "");
+                this.Close();
```
Comment (rammba) [inline WPFHospitalEditor/SearchResultDialog.xaml.cs:212]: This event is same as RoomRenovation.ScheduleRenovation method.

File: Schedule.API/Services/Procedures/RecommendationService.cs [hunk 7]
```diff
@@ -253,5 +254,94 @@ private IEnumerable<EquipmentRelocationDto> FindCommonEquipmentRelocationAppoint
...
+            IEnumerable<Shift> shiftsForSourceRoom = _shiftWrapper.Repository.GetShiftsByRoomID(dto.FirstRoomId);
+            foreach (Shift shift in shiftsForSourceRoom)
+            {
+                List<Shift> shiftList = new List<Shift>();
+                shiftList.Add(shift);
+                IEnumerable<RecommendationDto> recommendationsSourceRoom = FindRecommendationsInShifts(shiftList, 16);
+                if (recommendationsSourceRoom.Count() == 16)
+                {
+                    availableShiftsSource.Add(shift);
+                }
```
Comment (BoKna031) [inline Schedule.API/Services/Procedures/RecommendationService.cs:279]: What number 16 should represent

File: Schedule.API/Services/Procedures/RecommendationService.cs [hunk 8]
```diff
@@ -253,5 +254,94 @@ private IEnumerable<EquipmentRelocationDto> FindCommonEquipmentRelocationAppoint
...
+            }
+            foreach (Shift shift in shiftsForDestinationRoom)
+            {
+                List<Shift> shiftList = new List<Shift>();
+                shiftList.Add(shift);
+                IEnumerable<RecommendationDto> recommendationsDestinationRoom = FindRecommendationsInShifts(shiftList, 16);
+                if (recommendationsDestinationRoom.Count() == 16)
+                {
+                    availableShiftsDestination.Add(shift);
+                }
```
Comment (BoKna031) [inline Schedule.API/Services/Procedures/RecommendationService.cs:320]: Same as above xD

File: Schedule.API/Services/Procedures/RecommendationService.cs [hunk 9]
```diff
@@ -253,5 +254,94 @@ private IEnumerable<EquipmentRelocationDto> FindCommonEquipmentRelocationAppoint
...
+
+            List<TimeInterval> availableTimeIntervals = new List<TimeInterval>();
+            foreach (Shift shiftSource in availableShiftsSource)
+            {
+                foreach (Shift shiftDest in availableShiftsDestination)
+                {
+                    if (shiftDest.TimeInterval.Start == shiftSource.TimeInterval.Start && shiftDest.TimeInterval.End == shiftSource.TimeInterval.End)
+                    {
+                        availableTimeIntervals.Add(shiftDest.TimeInterval);
+                    }
```
Comment (BoKna031) [inline Schedule.API/Services/Procedures/RecommendationService.cs:331]: why do you look at strict boundaries

File: WPFHospitalEditor/AlternativeRenovationAppointments.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+namespace WPFHospitalEditor
+{
+    /// <summary>
+    /// Interaction logic for AlternativeRenovationAppointments.xaml
+    /// </summary>
+    public partial class AlternativeRenovationAppointments : Window
+    {
+
+        private readonly int roomId;
+        private readonly RoomRenovation rr;
```
Comment (BoKna031) [inline WPFHospitalEditor/AlternativeRenovationAppointments.xaml.cs:14]: rename it

File: WPFHospitalEditor/Controller/Interface/IRenovationServerController.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace WPFHospitalEditor.Controller.Interface
+{
+    public interface IRenovationServerController
+    {
+        string ScheduleRenovation(DateTime startDate, DateTime endDate, int doctorId, int patientId);
```
Comment (BoKna031) [inline WPFHospitalEditor/Controller/Interface/IRenovationServerController.cs:9]: you should use TimeInterval class for start and end date

File: WPFHospitalEditor/DTOs/EquipmentRelocationDto.cs [hunk 12]
```diff
@@ -9,5 +9,17 @@ public class EquipmentRelocationDto
         public int DestinationRoomId { get; set; }
         public int Amount { get; set; }
         public TimeInterval TimeInterval { get; set; }
+
+        public SchedulingDto toSchedulingDto()
+        {
+            SchedulingDto schedulingDto = new SchedulingDto()
+            {
+                FirstRoomId = SourceRoomId,
+                SecondRoomId = DestinationRoomId,
```
Comment (BoKna031) [inline WPFHospitalEditor/DTOs/EquipmentRelocationDto.cs:18]: What should first room and second room represents?

File: WPFHospitalEditor/DTOs/RenovationDto.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Text;
+using WPFHospitalEditor.Model;
+
+namespace WPFHospitalEditor.DTOs
+{
+    public class RenovationDto
+    {
+        public int FirstRoomId { get; set; }
+        public int SecondRoomId { get; set; }
```
Comment (BoKna031) [inline WPFHospitalEditor/DTOs/RenovationDto.cs:11]: ?

File: WPFHospitalEditor/DTOs/ScheduleRenovationDTO.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace WPFHospitalEditor.DTOs
+{
+    public class ScheduleRenovationDTO
+    {
+        public DateTime StartDate { get; internal set; }
+        public DateTime EndDate { get; internal set; }
```
Comment (BoKna031) [inline WPFHospitalEditor/DTOs/ScheduleRenovationDTO.cs:10]: Better use TimeInterval

Comment (BoKna031) [reply WPFHospitalEditor/DTOs/ScheduleRenovationDTO.cs:10]: If EF use this DTO then ignore previous comment

File: WPFHospitalEditor/DTOs/SchedulingDto.cs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+﻿using WPFHospitalEditor.Model;
+
+namespace WPFHospitalEditor.DTOs
+{
+    public class SchedulingDto
+    {
+        public int FirstRoomId { get; set; }
+        public int SecondRoomId { get; set; }
```
Comment (BoKna031) [inline WPFHospitalEditor/DTOs/SchedulingDto.cs:8]: Room naming

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    public partial class RoomRenovation : Window
+    {
+        private IRoomServerController roomServerController = new RoomServerController();
+        private IMapObjectController mapObjectController = new MapObjectController();
+        private IDoctorServerController doctorServerController = new DoctorServerController();
+        private IRenovationServerController renovationServerController = new RenovationServerController();
+
+        int mapObjectId;
+        DateTime startDate;
+        DateTime endDate;
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs:26]: TimeInterval

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        private void CloseClick(object sender, RoutedEventArgs e)
+        {
+            this.Close();
+        }
+
+        private void SetRoomsComboBox()
+        {
+            List<MapObject> mapObjects = mapObjectController.GetNeigbourMapObjects(mapObjectId);
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs:43]: rename it

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+            foreach (MapObject mo in mapObjects)
+            {
+                SecondRoomComboBox.Items.Add(mo.Id);
+            }
+        }
+
+        private void RenovateRoom(object sender, RoutedEventArgs e)
+        {
+            setSecondRoomId();
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs:52]: Destination room

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+            }
+            catch
+            {
+                MessageBox.Show("End time must be after start time!", "");
+            }
+        }
+
+        private void setSecondRoomId()
+        {
+            if (SecondRoomComboBox.SelectedIndex != 0)
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs:77]: Destination Room

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+            };
+            return renovationDto;
+        }
+
+        private void RenovationTypeSelectionChanged(object sender, SelectionChangedEventArgs e)
+        {
+            if(ComplexStackPanel != null)
+            {
+                if (RenovationTypeComboBox.SelectedIndex == 1)
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs:127]: Use Combobox.SelectedItem

File: WPFHospitalEditor/SearchResultDialog.xaml.cs [hunk 21]
```diff
@@ -22,6 +22,11 @@ public partial class SearchResultDialog : Window
         private readonly SearchType searchType;
         public static int selectedObjectId = -1;
         private Grid DynamicGrid;
+        private IDoctorServerController doctorServerController = new DoctorServerController();
+        private IExaminationServerController examinationServerController = new ExaminationServerController();
+        private IRenovationServerController renovationServerController = new RenovationServerController();
+        private DateTime startDate;
+        private DateTime endDate;
```
Comment (BoKna031) [inline WPFHospitalEditor/SearchResultDialog.xaml.cs:29]: Time interval

File: OcelotApiGateway/ocelot.global.json (hunk 1 above)
Comment (cakareli) [reply OcelotApiGateway/ocelot.global.json:294]: Resolved.

File: Schedule.API/Services/Procedures/RecommendationService.cs (hunk 2 above)
Comment (cakareli) [reply Schedule.API/Services/Procedures/RecommendationService.cs:274]: It should be like that, shiftList is used as parameter early created method, but I only need that one shift from shiftList.

File: Schedule.API/Services/Procedures/RecommendationService.cs (hunk 7 above)
Comment (cakareli) [reply Schedule.API/Services/Procedures/RecommendationService.cs:279]: It represents maximum number of appointments in one shift. It is defined as constant now.

File: Schedule.API/Services/Procedures/RecommendationService.cs (hunk 9 above)
Comment (cakareli) [reply Schedule.API/Services/Procedures/RecommendationService.cs:331]: I had to check if dates are same. It is fixed now.

File: WPFHospitalEditor/AlternativeRenovationAppointments.xaml.cs (hunk 10 above)
Comment (cakareli) [reply WPFHospitalEditor/AlternativeRenovationAppointments.xaml.cs:14]: Fixed.

File: WPFHospitalEditor/RoomRenovation.xaml.cs (hunk 17 above)
Comment (cakareli) [reply WPFHospitalEditor/RoomRenovation.xaml.cs:43]: Fixed

File: Schedule.API/DTOs/RoomRenovationDto.cs (hunk 4 above)
Comment (cakareli) [reply Schedule.API/DTOs/RoomRenovationDto.cs:8]: Fixed

File: WPFHospitalEditor/RoomRenovation.xaml.cs (hunk 16 above)
Comment (cakareli) [reply WPFHospitalEditor/RoomRenovation.xaml.cs:26]: Done

File: WPFHospitalEditor/Controller/Interface/IRenovationServerController.cs (hunk 11 above)
Comment (cakareli) [reply WPFHospitalEditor/Controller/Interface/IRenovationServerController.cs:9]: Done

File: WPFHospitalEditor/RoomRenovation.xaml.cs [hunk 22]
```diff
@@ -24,6 +24,7 @@ public partial class RoomRenovation : Window
         int mapObjectId;
         DateTime startDate;
         DateTime endDate;
```
Comment (BoKna031) [inline WPFHospitalEditor/RoomRenovation.xaml.cs]: delete this

**Review by rammba — APPROVED** [review]

**Pull request merged**